### PR TITLE
[Feature] [FEM] Erasing Elements Code

### PR DESCRIPTION
**Pull request by mac-the-bike**
Related Forum thread: https://forum.freecadweb.org/viewtopic.php?f=18&t=70978

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

File: src/Mod/Fem/App/FemSetElementNodesObject.cpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/* copy of: FemSetNodesObject.cpp */
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/App/FemSetElementNodesObject.cpp:3]: You are not Jürgen ;-). Please use your name here.

**Review by donovaly — CHANGES_REQUESTED** [review]
I was requested to review but at first I need to understand what the PR's aim is. Where is the forum discussion?
In general please link the forum discussion or provide a brief description when creating a PR.

Please also check the copyrights when you add a new file.

I cannot test your PR because it has merge conflicts. Can you please rebase the PR to current master?

**Comment by luzpaz** [comment]
@donovaly OP opened this forum thread: https://forum.freecadweb.org/viewtopic.php?f=18&t=70978

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7560">pipeline status</a> for feature branch [PR_7560](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7560). Pipeline [661855207](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/661855207) was triggered at [3e6c64f](https://github.com/FreeCAD/FreeCAD/pull/7560/commits/3e6c64f22f0493f2aab9d91fdda7be63e3243787). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by mac-the-bike** [comment]
th

**Comment by mac-the-bike** [comment]
I am sorry this reply has taken so long as I didn't expect replies here!

I am answering the questions in order, perhaps some have been resolved.

A - donovaly:
I have made these changes.

Question: can I just run git again, with the changes, and overwrite my current code in my current version?

B - luzpaz: 
I think you have done this.
[link]
https://forum.freecadweb.org/viewtopic.php?f=18&t=70978 
[/link]
I have also posted the "User's Guide" for the function at this link.

C -bernd:
I accidentally closed your comments, sorry.

D - Changes requested conflicts:
What do I have to do?
Do I have to change my code?
Do I copy the latest versions of AppFem.cpp and TaskAnalysisInfo.h and make my changes to THESE files?
Where are the latest versions if this is what is required?

This is what I am running:
[code]
OS: Ubuntu 22.04.1 LTS (GNOME-Flashback:GNOME/gnome-flashback-metacity)
Word size of FreeCAD: 64-bit
Version: 0.21.30159 (Git)
Build type: Unknown
Branch: master
Hash: 71015d23d051229db4b86d4e81dd887826c95d79
Python 3.10.6, Qt 5.15.3, Coin 4.0.0, Vtk 7.1.1, OCC 7.5.1
Locale: English/United Kingdom (en_GB)
[/code]

**Comment by luzpaz** [comment]
@mac-the-bike when you make changes to your code you can commit them and then push them to this branch. The PR will then reflect that you've made changes to the code and we can continue to review.

Take a look at https://www.youtube.com/watch?v=RGOj5yH7evk for more orientation. There is a timecode in the comments so you can skip around to relevant topics.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7560">pipeline status</a> for feature branch [PR_7560](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7560). Pipeline [679869536](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/679869536) was triggered at [3e6c64f](https://github.com/FreeCAD/FreeCAD/pull/7560/commits/3e6c64f22f0493f2aab9d91fdda7be63e3243787). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Base/BaseClass.cpp [hunk 2]
```diff
@@ -83,8 +83,16 @@ Type BaseClass::getTypeId() const
 void BaseClass::initSubclass(Base::Type &toInit,const char* ClassName, const char *ParentName,
                              Type::instantiationMethod method)
 {
+// ######  #####   #####    ####   #####
+// #       #    #  #    #  #    #  #    #
+// #####   #    #  #    #  #    #  #    #
+// #       #####   #####   #    #  #####
+// #       #   #   #   #   #    #  #   #
+// ######  #    #  #    #   ####   #    #
+// have commented out the next line as it always trips
```
Comment (donovaly) [inline src/Base/BaseClass.cpp:92]: Please only make changes that are relevant for your desired feature or fix. For further fixes, please make a separate PR.

File: src/Mod/Fem/App/FemSetElementNodesObject.cpp (hunk 1 above)
Comment (donovaly) [inline src/Mod/Fem/App/FemSetElementNodesObject.cpp:3]: You are not Jürgen. Please check the copyright statements

File: src/Mod/Fem/App/FemSetElementNodesObject.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
```
Comment (donovaly) [inline src/Mod/Fem/App/FemSetElementNodesObject.cpp:26]: This if code block is empty, therefore please omit it.

File: src/Mod/Fem/App/FemSetElementNodesObject.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+short FemSetElementNodesObject::mustExecute(void) const
+{
+    return 0;
+}
+
+PyObject *FemSetElementNodesObject::getPyObject()
+{
+    if (PythonObject.is(Py::_None())){
+        // ref counter is set to 1
```
Comment (donovaly) [inline src/Mod/Fem/App/FemSetElementNodesObject.cpp:57]: I don't understand this comment

File: src/Mod/Fem/App/FemSetElementNodesObject.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
+/* copy of: FemSetNodesObject.h */
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/App/FemSetElementNodesObject.h:3]: wrong copyright

File: src/Mod/Fem/Gui/CMakeLists.txt [hunk 6]
```diff
@@ -69,6 +69,7 @@ set(FemGui_UIC_SRCS
     DlgSettingsFemMystran.ui
     DlgSettingsFemZ88.ui
     TaskCreateNodeSet.ui
+    TaskCreateElementSet.ui
```
Comment (donovaly) [inline src/Mod/Fem/Gui/CMakeLists.txt:72]: Please keep the alphabetic sorting of the includes

File: src/Mod/Fem/Gui/Command.cpp [hunk 7]
```diff
@@ -60,8 +61,10 @@
 #include <Mod/Fem/App/FemPostPipeline.h>
 #endif
 
-
+using namespace FemGui;
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:64]: For what code do you need this namespace?

File: src/Mod/Fem/Gui/Command.cpp [hunk 8]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
 //================================================================================================
+
+/* start of Erase Element bit */
+DEF_STD_CMD_A(CmdFemDefineElementsSet);
+
+void DefineElementsCallback(void *ud, SoEventCallback *n)
+{
+    Fem::FemAnalysis* Analysis;
+
+//    Base::Console().Warning("DefineElemensCall\n");
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1143]: why this comment?

File: src/Mod/Fem/Gui/Command.cpp [hunk 9]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+
+    if (getConstraintPrerequisits(&Analysis))
+        return;
+
+    // show the wait cursor because this could take quite some time
+    Gui::WaitCursor wc;
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1149]: Is the wait cursor really necessary or just a paste from other FreeCAD code?

File: src/Mod/Fem/Gui/Command.cpp [hunk 10]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+
+    // When this callback function is invoked we must in either case leave the edit mode
+//    Gui::View3DInventorViewer* view = reinterpret_cast<Gui::View3DInventorViewer*>(n->getUserData());
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1152]: why this comment?

File: src/Mod/Fem/Gui/Command.cpp [hunk 11]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+    Gui::View3DInventorViewer* view = static_cast<Gui::View3DInventorViewer*>(n->getUserData());
+    view->setEditing(false);
+    view->removeEventCallback(SoMouseButtonEvent::getClassTypeId(), DefineElementsCallback, ud);
+    n->setHandled();
+
+    Gui::SelectionRole role;
+    std::vector<SbVec2f> clPoly = view->getGLPolygon(&role);
+    if (clPoly.size() < 3)
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1160]: Please comments so that other developers who might n e.g. 2 years need to  fix an issue, can understand what is done?
For you at the moment it is clear why the if is for < 3, but others in future would need this info

File: src/Mod/Fem/Gui/Command.cpp [hunk 12]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+        return;
+    if (clPoly.front() != clPoly.back())
+        clPoly.push_back(clPoly.front());
+
+    SoCamera* cam = view->getSoRenderManager()->getCamera();
+    SbViewVolume vv = cam->getViewVolume();
+    Gui::ViewVolumeProjection proj(vv);
+    Base::Polygon2d polygon;
+    for (std::vector<SbVec2f>::const_iterator it = clPoly.begin(); it != clPoly.end(); ++it)
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1169]: This is another example where comments would help others.
In general, add more comments (that's what I consistently tell everybody ;-) )

File: src/Mod/Fem/Gui/Command.cpp [hunk 13]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+        polygon.Add(Base::Vector2d((*it)[0], (*it)[1]));
+
+    std::vector<App::DocumentObject*> docObj = Gui::Selection().getObjectsOfType(Fem::FemMeshObject::getClassTypeId());
+    if (docObj.size() != 1)
+        return;
+
+//    const SMESHDS_Mesh* data = const_cast<SMESH_Mesh*>(static_cast<Fem::FemMeshObject*>(docObj[0])->FemMesh.getValue().getSMesh())->GetMeshDS();
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1176]: why this comment?

File: src/Mod/Fem/Gui/Command.cpp [hunk 14]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+
+    set << "[";
+    for (std::set<int>::const_iterator it = IntSet.begin(); it != IntSet.end(); ++it)
+        if (it == IntSet.begin())
+            set << *it;
+        else
+            set << "," << *it;
+    set << "]";
+
+//    Gui::Command::openCommand("Place robot");
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1201]: robot? In case you copied code from somewhere else, please review what is actually needed and omit not applicable code.
The name "robot" is not clear to me in the context of deleting mesh nodes.

File: src/Mod/Fem/Gui/Command.cpp [hunk 15]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+    sMenuText = QT_TR_NOOP("Element set by poly");
+    sToolTipText = QT_TR_NOOP("Create Element set by Poly");
+    sWhatsThis = "FEM_DefineElementsSet";
+    sStatusTip = QT_TR_NOOP("Create Element set by Poly");
+    sPixmap = "FEM_CreateElementsSet";
+}
+
+void CmdFemDefineElementsSet::activated(int)
+{
+//    Base::Console().Warning("Activated\n");
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1225]: why this comment?

File: src/Mod/Fem/Gui/Command.cpp [hunk 16]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+    std::vector<App::DocumentObject*> docObj = Gui::Selection().getObjectsOfType(Fem::FemMeshObject::getClassTypeId());
+
+    for (std::vector<App::DocumentObject*>::iterator it = docObj.begin(); it != docObj.end(); ++it) {
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1228]: please also here some comments what is done and why

File: src/Mod/Fem/Gui/Command.cpp [hunk 17]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+                return;
+            }
+        }
+
+    }
+}
+
+bool CmdFemDefineElementsSet::isActive()
+{
+//    Base::Console().Warning("isActive\n");
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1248]: why this comment?

File: src/Mod/Fem/Gui/Command.cpp [hunk 18]
```diff
@@ -1129,10 +1132,191 @@ bool CmdFemCreateNodesSet::isActive()
...
+    else {
+        QMessageBox::warning(Gui::getMainWindow(),
+            qApp->translate("CmdFemCreateElementsSet", "Wrong selection"),
+            qApp->translate("CmdFemCreateElementsSet", "Select a single FEM mesh or Elements set, please."));
+    }
+}
+
+bool CmdFemCreateElementsSet::isActive()
+{
+    //    Base::Console().Warning("ISACTIVE\n");
```
Comment (donovaly) [inline src/Mod/Fem/Gui/Command.cpp:1311]: comment?

File: src/Mod/Fem/Gui/TaskCreateElementSet.cpp [hunk 19]
```diff
@@ -0,0 +1,820 @@
+/***************************************************************************
+copy of TaskCreateNodeSet.cpp
+ ***************************************************************************/
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskCreateElementSet.cpp:5]: wrong copyright

File: src/Mod/Fem/Gui/TaskCreateElementSet.cpp [hunk 20]
```diff
@@ -0,0 +1,820 @@
...
+#include <SMESH_Mesh.hxx>
+#include <SMESHDS_Mesh.hxx>
+#include <SMDSAbs_ElementType.hxx>
+
+#endif
+
+#include <ui_TaskCreateElementSet.h>
+#include "TaskCreateElementSet.h"
+#include <QMessageBox>
+#include <Gui/Command.h>
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskCreateElementSet.cpp:52]: Please check what includes are actually necessary and please keep the alphabetic sorting:
App, Base, Gui, then the .h of the .cpp file then the rest.
See the other FEM files as reference

File: src/Mod/Fem/Gui/TaskCreateElementSet.cpp [hunk 21]
```diff
@@ -0,0 +1,820 @@
...
+#include <Mod/Fem/App/FemMesh.h>
+#include <Mod/Fem/App/FemMeshObject.h>
+#include <Mod/Fem/App/FemSetElementNodesObject.h>
+#include "ViewProviderFemMesh.h"
+#include "FemSelectionGate.h"
+using namespace FemGui;
+using namespace Gui;
+using namespace Fem;
+using namespace Base;
+using namespace boost;
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskCreateElementSet.cpp:77]: Pleas check what is actually needed. It seems you just copied everything here.

File: src/Mod/Fem/Gui/TaskCreateElementSet.cpp [hunk 22]
```diff
@@ -0,0 +1,820 @@
...
+int maxProject = 10;
+double **nodeCoords; // these are node coords
+int *nodeNumbers;    // these are node numbers - probably not required[100];
+
+void addFaceToMesh(const std::vector<const SMDS_MeshNode *> nodes, SMESHDS_Mesh *MeshDS, int EID)
+{
+    int nbNodes = nodes.size();
+    switch (nbNodes)
+    {
+    case 3:
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskCreateElementSet.cpp:106]: This is hardly readable. What is case 3 etc? Please use an enum instead. Then you get a self-explaining name like e.g. "case Nodes::Volume" of whatever the different cases stand for.

File: src/Mod/Fem/Gui/TaskCreateElementSet.cpp [hunk 23]
```diff
@@ -0,0 +1,820 @@
...
+
+void writeToFile(std::string fileName, SMESHDS_Mesh *newMesh, int *nodeNumbers,
+                 double **nodeCoords, int rows, int requiredType)
+{
+    /*
+     */
+
+    std::map<int, std::string> elType2D;
+    std::map<int, std::string> elType3D;
+    elType2D[3] = "S3";
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskCreateElementSet.cpp:291]: Please use en enum here since numbers are hard to get later in the code. Also please keep the entries sorted for better overview

File: src/Mod/Fem/Gui/TaskCreateElementSet.cpp [hunk 24]
```diff
@@ -0,0 +1,820 @@
...
+        fprintf(fptr, "%s", "Efaces\n");
+    }
+
+    fclose(fptr);
+    return;
+}
+
+void printData(const SMESHDS_Mesh *srcMesh)
+{
+    //     Base::Console().Warning("PRINT START\n");
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskCreateElementSet.cpp:392]: comment?

File: src/Mod/Fem/Gui/TaskCreateElementSet.h [hunk 25]
```diff
@@ -0,0 +1,98 @@
+/***************************************************************************
+copy of TaskCreateNodeSet.h bit with createElementSet
+ ***************************************************************************/
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskCreateElementSet.h:5]: wrong copyright

File: src/Mod/Fem/Gui/TaskDlgCreateElementSet.cpp [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskDlgCreateElementSet.cpp:26]: empty if -> omit it

File: src/Mod/Fem/Gui/TaskDlgCreateElementSet.cpp [hunk 27]
```diff
@@ -0,0 +1,110 @@
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskDlgCreateElementSet.cpp:2]: copyright

File: src/Mod/Fem/Gui/TaskDlgCreateElementSet.h [hunk 28]
```diff
@@ -0,0 +1,74 @@
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskDlgCreateElementSet.h:2]: copyright

File: src/Mod/Fem/Gui/ViewProviderSetElementNodes.cpp [hunk 29]
```diff
@@ -0,0 +1,57 @@
+/* copy of ./src/Mod/Fem/Gui/ViewProviderSetNodes.cpp */
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/ViewProviderSetElementNodes.cpp:3]: copyright

File: src/Mod/Fem/Gui/ViewProviderSetElementNodes.cpp [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
```
Comment (donovaly) [inline src/Mod/Fem/Gui/ViewProviderSetElementNodes.cpp:27]: empty if

File: src/Mod/Fem/Gui/ViewProviderSetElementNodes.h [hunk 31]
```diff
@@ -0,0 +1,47 @@
+/***************************************************************************
+ *   Copyright (c) 2013 Jürgen Riegel (FreeCAD@juergen-riegel.net)         *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/ViewProviderSetElementNodes.h:2]: copyright

**Review by donovaly — CHANGES_REQUESTED** [review]
Many thanks for your PR.
I have several comments.

In general add more comments to help others to understand the code.

There are some areas in your code where it looks as if you pasted in some code from other parts of FreeCAD. This is OK, but please check what of the code is actually necessary. Omit every unused include and commented out code.

**Comment by luzpaz** [comment]
bump @mac-the-bike

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7560">pipeline status</a> for feature branch [PR_7560](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7560). Pipeline [761702984](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/761702984) was triggered at [3e6c64f](https://github.com/FreeCAD/FreeCAD/pull/7560/commits/3e6c64f22f0493f2aab9d91fdda7be63e3243787). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by sliptonic** [comment]
The PR has review comments that haven't been resolved and also has current merge conflicts.  I'm moving the PR to draft status.  @mac-the-bike if you want to move forward, please address the review comments and rebase the pr.  Then move it back out of draft status.

**Comment by mac-the-bike** [comment]
sliptonic: The PR review comments:

I have resolved all the comments because donovaly didn't make any new comments after Oct 30th

After resolving I created a new version of the code, (because I don't know how to overwrite the original version), and I called it version 3, pull request 7769, as below:

https://github.com/FreeCAD/FreeCAD/compare/main...mac-the-bike:FreeCAD:erase-version-3

this is the code that should be merged into FREECAD.

**Comment by marioalexis84** [comment]
> @marioalexis84 would you have time to take a look at this ?

Of course, when @mac-the-bike has created the pull request.
Some fast comments:
* The curls in function definitions should be in a new line. So change:
```
void f() {
...
}
```
to
```
void f()
{
...
}
```
* The conditional blocks should use curls even if they have only one line.
So change:
```
if (condition)
    f();
else
    g();
```
to
```
if (condition) {
    f();
}
else {
    g();
}
```
* Use range-based loops instead of iterators when is possible. 

There is a pre-commit script for code formatting which is used currently. Here is the explanation: https://freecad.github.io/DevelopersHandbook/gettingstarted/


I think there is no need to ammend the licence with each change in the some file, and according to wmayer's suggestion, for the new files it should be like this: https://github.com/FreeCAD/FreeCAD/blob/main/src/Gui/TaskView/TaskOrientation.cpp

**Comment by mac-the-bike** [comment]
sliptonic

it's been some time, but...
i have generated a PR, 11492, that is the latest version of:
-- FreeCAD, as of November 18
-- modified code, 9 files
-- new code, 10 files

Can you do me a favour and make sure I have completed all the steps to get this code into FreeCAD? 
If it incomplete, then e-mail me, 

Thanks.
